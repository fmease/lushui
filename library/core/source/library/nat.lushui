;; Natural numbers and related functions.
module

use topmost.type.Type
use topmost.(option.Option bool.Bool text.Text)

@public @intrinsic
data Nat: Type

@public
use add as +

@public @intrinsic
add: Nat -> Nat -> Nat

@public
use subtract as -

@public @intrinsic
subtract: Nat -> Nat -> Option Nat

@public
use panicking-subtract as -?

;;; @Temporary
@public @intrinsic
panicking-subtract: Nat -> Nat -> Nat

@public
use multiply as *

@public @intrinsic
multiply: Nat -> Nat -> Nat

@public
use divide as /

@public @intrinsic
divide: Nat -> Nat -> Option Nat

@public
use equal as ==

@public @intrinsic
equal: Nat -> Nat -> Bool

;;; equal (n m: Nat): Bool =
;;;     case duple Nat Nat n m of
;;;         duple Nat Nat \n n => true
;;;         \_ => false

@public
use less as <

@public @intrinsic
less: Nat -> Nat -> Bool

@public
use less-equal as =<

@public @intrinsic
less-equal: Nat -> Nat -> Bool

@public
use greater as >

@public @intrinsic
greater: Nat -> Nat -> Bool

@public
use greater-equal as >=

@public @intrinsic
greater-equal: Nat -> Nat -> Bool

@public @intrinsic
display: Nat -> Text

use topmost.(bool ordering)

@public
compare (n: Nat) (m: Nat): ordering.Ordering =
    case equal n m of
        bool.true => ordering.equal
        bool.false => case less n m of
            bool.true => ordering.less
            bool.false => ordering.greater

;;; @public @intrinsic
;;; to-int: Nat -> topmost.int.Int
