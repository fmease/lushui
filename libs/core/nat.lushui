use super.(option.Option bool.Bool text.Text)

@public @foreign
data Nat: Type

@public
use add as +

@public @foreign
add: Nat -> Nat -> Nat

@public
use subtract as -

@public @foreign
subtract: Nat -> Nat -> Option Nat

@public
use panicking-subtract as -?

;; @Temporary
@public @foreign
panicking-subtract: Nat -> Nat -> Nat

@public
use multiply as *

@public @foreign
multiply: Nat -> Nat -> Nat

@public
use divide as /

@public @foreign
divide: Nat -> Nat -> Option Nat

@public
use equal as ==

@public @foreign
equal: Nat -> Nat -> Bool

;; equal (n m: Nat): Bool =
;;     case duple Nat Nat n m of
;;         duple Nat Nat \n n => true
;;         \_ => false

@public
use less as <

@public @foreign
less: Nat -> Nat -> Bool

@public
use less-equal as =<

@public @foreign
less-equal: Nat -> Nat -> Bool

@public
use greater as >

@public @foreign
greater: Nat -> Nat -> Bool

@public
use greater-equal as >=

@public @foreign
greater-equal: Nat -> Nat -> Bool

@public @foreign
display: Nat -> Text

@public
use super.(bool ordering)

@public
compare (n m: Nat): ordering.Ordering =
    case equal n m of
        bool.true => ordering.equal
        bool.false => case less n m of
            bool.true => ordering.less
            bool.false => ordering.greater

;; @public @foreign
;; to-int: Nat -> Nat -> core.int.Int
