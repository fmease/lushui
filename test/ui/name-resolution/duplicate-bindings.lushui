use extern.core.nat.Nat

data Alpha of
    alpha (beta: Alpha): Alpha

Alpha = Type

beta: Nat = 1000

module pod of
    use super.Nat

    data Nat of
        zero: Nat
        successor: Nat -> Nat

beta (x: Nat) (y: Nat): Nat = y

pod: Nat = beta

data ALIKE of
    ALIKE: ALIKE

zero: Nat = 0

use pod.zero
