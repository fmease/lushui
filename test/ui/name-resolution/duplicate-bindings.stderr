error[E020]: ‘Alpha’ is defined multiple times in this scope
  ┌─ ${TEST_FOLDER}/name-resolution/duplicate-bindings.lushui:3:6
  │
3 │ data Alpha: Type of
  │      ═════ conflicting definition
  │
  ├─ ${TEST_FOLDER}/name-resolution/duplicate-bindings.lushui:6:1
  │
6 │ Alpha: Type = Type
  │ ═════ conflicting definition

error[E020]: ‘beta’ is defined multiple times in this scope
   ┌─ ${TEST_FOLDER}/name-resolution/duplicate-bindings.lushui:8:1
   │
 8 │ beta: Nat = 1000
   │ ════ conflicting definition
   │
   ├─ ${TEST_FOLDER}/name-resolution/duplicate-bindings.lushui:17:1
   │
17 │ beta (x: Nat) (y: Nat): Nat = y
   │ ════ conflicting definition

error[E020]: ‘pod’ is defined multiple times in this scope
   ┌─ ${TEST_FOLDER}/name-resolution/duplicate-bindings.lushui:10:8
   │
10 │ module pod of
   │        ═══ conflicting definition
   │
   ├─ ${TEST_FOLDER}/name-resolution/duplicate-bindings.lushui:19:1
   │
19 │ pod: Nat = beta
   │ ═══ conflicting definition

error[E020]: ‘Nat’ is defined multiple times in this scope
   ┌─ ${TEST_FOLDER}/name-resolution/duplicate-bindings.lushui:11:15
   │
11 │     use super.Nat
   │               ═══ conflicting definition
   │
   ├─ ${TEST_FOLDER}/name-resolution/duplicate-bindings.lushui:13:10
   │
13 │     data Nat: Type of
   │          ═══ conflicting definition

error[E020]: ‘zero’ is defined multiple times in this scope
   ┌─ ${TEST_FOLDER}/name-resolution/duplicate-bindings.lushui:24:1
   │
24 │ zero: Nat = 0
   │ ════ conflicting definition
   │
   ├─ ${TEST_FOLDER}/name-resolution/duplicate-bindings.lushui:26:9
   │
26 │ use pod.zero
   │         ════ conflicting definition

error: aborting due to 5 previous errors

